Services/communication_service.c

```diff
@@ -48,18 +60,156 @@ SAT_returnState communication_service_app(csp_packet_t *packet) {
         return SATR_ERROR;
       }
+      break;
 
+    case GET_CONTROL:
+      HAL_S_getControl (&S_PA.PA_status, &S_PA.PA_mode);
+      S_PA.PA_status = csp_hton32(S_PA.PA_status);
+      S_PA.PA_mode = csp_hton32(S_PA.PA_mode);
+      status = 0;
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_PA.PA_status, sizeof(uint32_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + sizeof(uint32_t)], &S_PA.PA_mode, sizeof(uint32_t));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + sizeof(uint32_t) + 1);
+
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
       break;
 
+    case GET_ENCODER:
+      HAL_S_getEncoder(&S_enc.Enc_scrambler, &S_enc.Enc_filter, &S_enc.Enc_modulation, &S_enc.Enc_rate);
+      S_enc.Enc_scrambler = csp_hton32(S_enc.Enc_scrambler);
+      S_enc.Enc_filter = csp_hton32(S_enc.Enc_filter);
+      S_enc.Enc_modulation = csp_hton32(S_enc.Enc_modulation);
+      S_enc.Enc_rate = csp_hton32(S_enc.Enc_rate);
+      status = 0;
+
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_enc.Enc_scrambler, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + sizeof(int8_t)], &S_enc.Enc_filter, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + 2*sizeof(int8_t)], &S_enc.Enc_modulation, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE + 3*sizeof(int8_t)], &S_enc.Enc_rate, sizeof(int8_t));
+
+      set_packet_length(packet, sizeof(int8_t) + 4*sizeof(int8_t) + 1);
+
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+      break;
+
+    case GET_PApower:
+      HAL_S_getpaPower(&S_config.S_paPower);
+      S_config.S_paPower = csp_hton32(S_config.S_paPower);
+      status = 0;
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_config.S_paPower, sizeof(uint32_t));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(uint32_t) + 1);
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+      break;
+
+    case GET_FirmwareV:
+        HAL_S_getFirmwareV(&S_Firmware.Version);
+        S_Firmware.Version = csp_hton32(S_Firmware.Version);// check if 32 is correct/necessary
+        status = 0;
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_Firmware.Version, sizeof(float));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(float) + 1);
+        if (queue_response(packet) != SATR_OK) {
+          return SATR_ERROR;
+        }
+        break;
+
+    case GET_STATUS:
+      HAL_S_getStatus (&S_status.PWRGD, &S_status.TXL);
+      S_status.PWRGD = csp_hton32(S_status.PWRGD);
+      S_status.TXL = csp_hton32(S_status.TXL);
+      status = 0;
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
```

Comment:
Just check the size of the structs that you're going to copy into the packet so we don't get segmentation faults, use csp_buffer_data_size()

Reply:
something like 
```
if (sizeof(S_status) > csp_buffer_data_size() - 1) { // minus one for subservice
 // log, return error
}
```

Reply:
Would that minus one be different based on the subservice (having arguments, size of the structure, etc.)?

Reply:
Only if something else is going into the packet data. Perhaps a better way to think about it would be `sizeof(S_status) + 1 > csp_buffer_data_size()`, since it's the return data we're concerned with. Basically, we'll need to account for anything we're about to `memcpy` into that memory space